### PR TITLE
Fix critical bugs

### PR DESCRIPTION
**Pull request by szepeviktor**
### Description

The clone of that fruitful PR in the packaging repository.
https://github.com/co-cart/co-cart/pull/388

### Types of changes

Contains critical bug fixes,
and the type fixes **leading up** to finding those bugs during static analysis.

### Checklist:
- [ ] My code is tested
- [ ] My code has proper inline documentation 
- [ ] I've included any necessary tests 
- [ ] I've included developer documentation 
- [ ] I've added proper labels to this pull request 

I hope you take fancy to run PHPStan as there are sure benefits.

### CODE REVIEW

File: includes/classes/admin/class-cocart-wc-admin-system-status.php [hunk 1]
```diff
@@ -45,8 +45,8 @@ public function __construct() {
 		add_filter( 'woocommerce_debug_tools', array( $this, 'debug_buttons' ) );
 
 		// Add tools to REST System Status tool.
-		add_filter( 'woocommerce_rest_insert_system_status_tool', array( $this, 'maybe_verify_database' ), 10, 2 );
-		add_filter( 'woocommerce_rest_insert_system_status_tool', array( $this, 'maybe_update_database' ), 10, 2 );
+		add_filter( 'woocommerce_rest_insert_system_status_tool', array( $this, 'maybe_verify_database' ) );
+		add_filter( 'woocommerce_rest_insert_system_status_tool', array( $this, 'maybe_update_database' ) );
```
Comment (szepeviktor) [inline includes/classes/admin/class-cocart-wc-admin-system-status.php:49]: These `maybe_*_database` functions do not accept a second parameter.

File: includes/cocart-formatting-functions.php [hunk 2]
```diff
@@ -274,7 +274,7 @@ function cocart_format_variation_data( $attributes, $product ) {
 		if ( taxonomy_exists( $taxonomy ) ) {
 			// If this is a term slug, get the term's nice name.
 			$term = get_term_by( 'slug', $value, $taxonomy );
-			if ( ! is_wp_error( $term ) && $term && $term->name ) {
+			if ( $term && $term->name ) {
```
Comment (szepeviktor) [inline includes/cocart-formatting-functions.php:277]: `get_term_by` does not return `WP_Error`.

File: includes/cocart-rest-functions.php [hunk 3]
```diff
@@ -262,7 +262,7 @@ function cocart_set_uploaded_image_as_attachment( $upload, $id = 0 ) {
 		'post_content'   => $content,
 	);
 
-	$attachment_id = wp_insert_attachment( $attachment, $upload['file'], $id );
+	$attachment_id = wp_insert_attachment( $attachment, $upload['file'], $id, true );
```
Comment (szepeviktor) [inline includes/cocart-rest-functions.php:265]: Make `wp_insert_attachment` return `WP_Error`.

File: uninstall.php [hunk 4]
```diff
@@ -45,7 +45,7 @@
 	require_once dirname( __FILE__ ) . '/includes/classes/class-cocart-helpers.php';
 
 	// Delete WooCommerce Admin Notes.
-	if ( CoCart\Helpers::is_wc_admin_enabled() ) {
+	if ( CoCart\Help::is_wc_admin_enabled() ) {
```
Comment (szepeviktor) [inline uninstall.php:48]: A missing rename.

File: includes/classes/class-cocart-install.php [hunk 5]
```diff
@@ -570,10 +570,12 @@ public static function redirect_getting_started( $plugin ) {
...
 		 */
 		if ( defined( 'WP_CLI' ) && WP_CLI ) {
+			/* translators: %1$s: CoCart */
+			$message = sprintf( __( 'Get started with %1$s here:', 'cart-rest-api-for-woocommerce' ), 'CoCart' );
 			\WP_CLI::log(
 				\WP_CLI::colorize(
-					/* translators: %1$s: message, %2$s: URL, %3$s: CoCart */
-					'%y' . sprintf( '🎉 %1$s %2$s', __( 'Get started with %3$s here:', 'cart-rest-api-for-woocommerce' ), $getting_started, 'CoCart' ) . '%n'
+					/* translators: %1$s: message, %2$s: URL */
+					'%y' . sprintf( '🎉 %1$s %2$s', $message, $getting_started ) . '%n'
```
Comment (szepeviktor) [inline includes/classes/class-cocart-install.php:578]: `%3` is no more.
🌪️

**Comment by seb86** [comment]
@szepeviktor Thanks again for submitting this PR. Does PHPStan automatically apply these corrections or only notifies you on what needs to be fixed? Also are any changes related to a specific WordPress version for compatibility?

**Comment by szepeviktor** [comment]
> Does PHPStan automatically apply these corrections?

No. PHPStan has a learning curve.
Basically this is the rule of thumb: You write code PHPStan understands ✔️
PHPStan shows a false positive 👉🏻 Write code PHPStan understands!

> Also are any changes related to a specific WordPress version for compatibility?

No, these are generic fixes. This is PHPStan level 0.

**Comment by szepeviktor** [comment]
I want you to be a PHPStan fan.

**Comment by seb86** [comment]
@szepeviktor Have I done the configuration wrong or am I missing something? 
image

**Comment by szepeviktor** [comment]
`phpstan.neon`

```neon
# composer require php-stubs/woocommerce-stubs php-stubs/wp-cli-stubs szepeviktor/phpstan-wordpress --dev

includes:
    - vendor/szepeviktor/phpstan-wordpress/extension.neon
parameters:
    paths:
        - includes/
        - uninstall.php
    excludePaths:
        - includes/cocart-deprecated-functions.php
    bootstrapFiles:
        - vendor/php-stubs/woocommerce-stubs/woocommerce-stubs.php
    scanFiles:
        - vendor/php-stubs/wp-cli-stubs/wp-cli-stubs.php
        - includes/cocart-deprecated-functions.php
        # For analyzing cocart-core only
        - packages/admin/includes/class-cocart-admin-notices.php
        - packages/products-api/includes/class-cocart-datetime.php
    level: 0
```

If you bump into errors try debugging: `vendor/bin/phpstan analyze --debug`

**Comment by seb86** [comment]
Thanks. Got it working. Crazy how much memory it needs to work.

**Comment by szepeviktor** [comment]
PHPStan has 100 developers built in!

👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻
👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻👨‍💻

**Pull request merged**